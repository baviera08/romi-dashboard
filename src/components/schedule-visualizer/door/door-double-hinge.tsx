import React from 'react';
import doorStyles from './door-style';
import { DoorProps, getDoorStyle } from './door';

/**
 * Double hinge doors:
 * - hinges are located at both (v1_x, v1_y) and (v2_x, v2_y)
 * - motion range = door swing ranges in DEGREES (assume symmetric)
 * - same motion-direction selection as single hinge
 */
const DoubleHingeDoor = React.forwardRef(function(
  props: DoorProps,
  ref: React.Ref<SVGGElement>,
): React.ReactElement {
  const { door, onClick, currentMode } = props;
  const { v1_x, v1_y, v2_x, v2_y } = door;
  const [hingeX1, hingeY1, hingeX2, hingeY2] = [v1_x, v1_y, v2_x, v2_y];
  const [extendX1, extendY1] = [hingeX1 + (v2_x - v1_x) / 2, hingeY1 + (v2_y - v1_y) / 2];
  const classes = doorStyles();
  const doorStyle = getDoorStyle(classes, currentMode);
  return (
    <>
      <g ref={ref} onClick={e => onClick && onClick(e as any, door)}>
        <line
          className={`${classes.doorMarker} ${classes.door} ${doorStyle}`}
          x1={hingeX1}
          y1={-hingeY1}
          x2={extendX1}
          y2={-extendY1}
        />
      </g>
      <g ref={ref} onClick={e => onClick && onClick(e as any, door)}>
        <line
          className={`${classes.doorMarker} ${classes.door} ${doorStyle}`}
          x1={extendX1}
          y1={-extendY1}
          x2={hingeX2}
          y2={-hingeY2}
        />
      </g>
    </>
  );
});

export default DoubleHingeDoor;
