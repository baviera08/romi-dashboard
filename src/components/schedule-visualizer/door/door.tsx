import * as RomiCore from '@osrf/romi-js-core-interfaces';
import React from 'react';
import { Spinner } from '../spinner';
import doorStyles from './door-style';
import SingleHingeDoor from './door-single-hinge';
import DoubleHingeDoor from './door-double-hinge';

export enum DoorMode {
  CLOSE,
  PROCESS,
  OPEN,
}

export enum DoorType {
  UNDEFINED,
  SINGLE_SLIDING,
  DOUBLE_SLIDING,
  SINGLE_TELESCOPE,
  DOUBLE_TELESCOPE,
  SINGLE_SWING,
  DOUBLE_SWING,
}

/**
 * currentMode: current mode of the door. E.g: 0 = CLOSE
 * door: door information on map
 * doorState: current state of the door.
 * onClick: action to trigger on click.
 */
export interface DoorProps {
  currentMode?: number;
  door: RomiCore.Door;
  doorState?: RomiCore.DoorState;
  onClick?(e: React.MouseEvent<SVGGElement>, door: RomiCore.Door): void;
}

export const getDoorStyle = (classes: any, currentMode: number | undefined) => {
  switch (currentMode) {
    case DoorMode.OPEN:
      return classes.doorOpen;
    case DoorMode.PROCESS:
      return classes.doorProcess;
    default:
      return classes.doorClose;
  }
};

const Door = React.forwardRef(function(
  props: DoorProps,
  ref: React.Ref<SVGGElement>,
): React.ReactElement {
  const { door, onClick, currentMode } = props;
  const { door_type: doorType } = door;
  return (
    <>
      {doorType === DoorType.SINGLE_SLIDING && (
        <SingleHingeDoor door={door} onClick={onClick} currentMode={currentMode} />
      )}
      {doorType === DoorType.DOUBLE_SLIDING && (
        <DoubleHingeDoor door={door} onClick={onClick} currentMode={currentMode} />
      )}
      {currentMode === DoorMode.PROCESS && (
        <g>
          <Spinner cx={door.v1_x} cy={-door.v1_y} r={0.4} strokeWidth={0.1}></Spinner>
        </g>
      )}
    </>
  );
});

export default Door;
