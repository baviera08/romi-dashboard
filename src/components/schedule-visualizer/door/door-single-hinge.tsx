import React from 'react';
import doorStyles from './door-style';
import { DoorProps, getDoorStyle } from './door';

/*
 * Single hinge doors:
 *  - hinge is located at (v1_x, v1_y)
 *  - door extends till (v2_x, v2_y)
 *  - motion_range = door swing range in DEGREES
 *  - there are two possible motions: clockwise and anti-clockwise
 *  - selected by the motion_direction parameter, which is +1 or -1
 */
const SingleHingeDoor = React.forwardRef(function(
  props: DoorProps,
  ref: React.Ref<SVGGElement>,
): React.ReactElement {
  const { door, onClick, currentMode } = props;
  const { v1_x, v1_y, v2_x, v2_y } = door;
  const [hingeX, hingeY, extendX, extendY] = [v1_x, v1_y, v2_x, v2_y];
  const classes = doorStyles();

  return (
    <>
      <g>
        <line
          className={`${classes.doorMarker} ${classes.door} ${getDoorStyle(classes, currentMode)}`}
          x1={hingeX}
          y1={-hingeY}
          x2={extendX}
          y2={-extendY}
        />
      </g>
      <g ref={ref} onClick={e => onClick && onClick(e as any, door)}>
        <line
          className={`${classes.doorMarker} ${classes.door} ${classes.doorTransparent}`}
          x1={hingeX}
          y1={-hingeY}
          x2={extendX}
          y2={-extendY}
        />
      </g>
    </>
  );
});

export default SingleHingeDoor;
