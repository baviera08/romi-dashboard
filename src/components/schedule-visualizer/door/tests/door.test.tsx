import * as RomiCore from '@osrf/romi-js-core-interfaces';
import { createMount } from '@material-ui/core/test-utils';
import { ReactWrapper } from 'enzyme';
import React from 'react';
import SingleHingeDoor from '../door-single-hinge';
import SingleSlideDoor from '../door-single-slide';
import DefaultDoor from '../door-default';
import DoubleHingeDoor from '../door-double-hinge';
import { DoorProps, getDoorStyle } from '../door';
import DoubleSlideDoor from '../door-double-slide';
import doorStyles from '../door-style';

const mainDoor = {
  name: 'main_door',
  v1_x: 8.2,
  v1_y: -5.5,
  v2_x: 7.85,
  v2_y: -6.2,
  door_type: 2,
  motion_range: -1.571,
  motion_direction: 1,
};

const buildDoorWrapper = (
  Component: (props: DoorProps) => React.ReactElement,
  mainDoor: RomiCore.Door,
  currentMode: number,
) => {
  return mount(
    <svg>
      <Component
        v1={[mainDoor.v1_x, mainDoor.v1_y]}
        v2={[mainDoor.v2_x, mainDoor.v2_y]}
        door={mainDoor}
        currentMode={currentMode}
      />
    </svg>,
  );
};

const mount = createMount();

const getSingleDoorSVGLine = (wrapper: ReactWrapper) => {
  return wrapper
    .find(DefaultDoor)
    .at(0)
    .find('line')
    .at(0);
};

test('Trigger click event', async () => {
  let clicked = false;
  const handleClick = () => {
    clicked = true;
  };
  const wrapper = mount(
    <svg>
      <DefaultDoor
        v1={[mainDoor.v1_x, mainDoor.v1_y]}
        v2={[mainDoor.v2_x, mainDoor.v2_y]}
        onClick={handleClick}
      />
    </svg>,
  );
  wrapper
    .find(DefaultDoor)
    .at(0)
    .find('g')
    .simulate('click');
  expect(clicked).toBe(true);

  wrapper.unmount();
});

test('Get correct door style', () => {
  const doorStyle = {
    doorOpen: 'doorOpenStyle',
    doorClose: 'doorCloseStyle',
    doorProcess: 'doorProcessStyle',
  };
  const open = getDoorStyle(doorStyle, RomiCore.DoorMode.MODE_OPEN);
  expect(open).toEqual(doorStyle.doorOpen);
  const closed = getDoorStyle(doorStyle, RomiCore.DoorMode.MODE_CLOSED);
  expect(closed).toEqual(doorStyle.doorClose);
  const moving = getDoorStyle(doorStyle, RomiCore.DoorMode.MODE_MOVING);
  expect(moving).toEqual(doorStyle.doorProcess);
});

describe('Checks assignation of styles on different states of SingleHingeDoor', () => {
  test('Style of SingleHingeDoor on Open', () => {
    const wrapper = buildDoorWrapper(SingleHingeDoor, mainDoor, RomiCore.DoorMode.MODE_OPEN);
    const singleDoorSVGLine = getSingleDoorSVGLine(wrapper);
    expect(singleDoorSVGLine.hasClass(/(makeStyles)-(doorOpen)-(\d+)/)).toBe(true);
    wrapper.unmount();
  });

  test('Style of SingleHingeDoor on Close', () => {
    const wrapper = buildDoorWrapper(SingleHingeDoor, mainDoor, RomiCore.DoorMode.MODE_CLOSED);
    const singleDoorSVGLine = getSingleDoorSVGLine(wrapper);
    expect(singleDoorSVGLine.hasClass(/(makeStyles)-(doorClose)-(\d+)/)).toBe(true);
    wrapper.unmount();
  });

  test('Style of SingleHingeDoor on Process', () => {
    const wrapper = buildDoorWrapper(SingleHingeDoor, mainDoor, RomiCore.DoorMode.MODE_MOVING);
    const singleDoorSVGLine = getSingleDoorSVGLine(wrapper);
    expect(singleDoorSVGLine.hasClass(/(makeStyles)-(doorProcess)-(\d+)/)).toBe(true);
    wrapper.unmount();
  });
});

describe('Checks assignation of styles on different states', () => {
  test('Style of SingleSlideDoor on Open', () => {
    const wrapper = buildDoorWrapper(SingleSlideDoor, mainDoor, RomiCore.DoorMode.MODE_OPEN);
    const singleDoorSVGLine = getSingleDoorSVGLine(wrapper);
    expect(singleDoorSVGLine.hasClass(/(makeStyles)-(doorOpen)-(\d+)/)).toBe(true);
    wrapper.unmount();
  });

  test('Style of SingleSlideDoor on Close', () => {
    const wrapper = buildDoorWrapper(SingleSlideDoor, mainDoor, RomiCore.DoorMode.MODE_CLOSED);
    const singleDoorSVGLine = getSingleDoorSVGLine(wrapper);
    expect(singleDoorSVGLine.hasClass(/(makeStyles)-(doorClose)-(\d+)/)).toBe(true);
    wrapper.unmount();
  });

  test('Style of SingleSlideDoor on Process', () => {
    const wrapper = buildDoorWrapper(SingleSlideDoor, mainDoor, RomiCore.DoorMode.MODE_MOVING);
    const singleDoorSVGLine = getSingleDoorSVGLine(wrapper);
    expect(singleDoorSVGLine.hasClass(/(makeStyles)-(doorProcess)-(\d+)/)).toBe(true);
    wrapper.unmount();
  });
});

describe('Checks assignation of styles on DobleHingeDoors', () => {
  const getSingleDoorSVGLine = (wrapper: ReactWrapper) => {
    return wrapper
      .find(DefaultDoor)
      .at(0)
      .find('line')
      .at(0);
  };

  test('Style of DoubleHingeDoors on Open', () => {
    const wrapper = buildDoorWrapper(DoubleHingeDoor, mainDoor, RomiCore.DoorMode.MODE_OPEN);
    const singleDoorSVGLine = getSingleDoorSVGLine(wrapper);
    expect(singleDoorSVGLine.hasClass(/(makeStyles)-(doorOpen)-(\d+)/)).toBe(true);
    wrapper.unmount();
  });

  test('Style of DoubleHingeDoors on Close', () => {
    const wrapper = buildDoorWrapper(DoubleHingeDoor, mainDoor, RomiCore.DoorMode.MODE_CLOSED);
    const singleDoorSVGLine = getSingleDoorSVGLine(wrapper);
    expect(singleDoorSVGLine.hasClass(/(makeStyles)-(doorClose)-(\d+)/)).toBe(true);
    wrapper.unmount();
  });

  test('Style of DoubleHingeDoors on Process', () => {
    const wrapper = buildDoorWrapper(DoubleHingeDoor, mainDoor, RomiCore.DoorMode.MODE_MOVING);
    const singleDoorSVGLine = getSingleDoorSVGLine(wrapper);
    expect(singleDoorSVGLine.hasClass(/(makeStyles)-(doorProcess)-(\d+)/)).toBe(true);
    wrapper.unmount();
  });
});

describe('Checks assignation of styles on DoubleSlideDoor', () => {
  const getSingleDoorSVGLine = (wrapper: ReactWrapper) => {
    return wrapper
      .find(DefaultDoor)
      .at(0)
      .find('line')
      .at(0);
  };

  test('Style of DoubleSlideDoor on Open', () => {
    const wrapper = buildDoorWrapper(DoubleSlideDoor, mainDoor, RomiCore.DoorMode.MODE_OPEN);
    const singleDoorSVGLine = getSingleDoorSVGLine(wrapper);
    expect(singleDoorSVGLine.hasClass(/(makeStyles)-(doorOpen)-(\d+)/)).toBe(true);
    wrapper.unmount();
  });

  test('Style of DoubleSlideDoor on Close', () => {
    const wrapper = buildDoorWrapper(DoubleSlideDoor, mainDoor, RomiCore.DoorMode.MODE_CLOSED);
    const singleDoorSVGLine = getSingleDoorSVGLine(wrapper);
    expect(singleDoorSVGLine.hasClass(/(makeStyles)-(doorClose)-(\d+)/)).toBe(true);
    wrapper.unmount();
  });

  test('Style of DoubleSlideDoor on Process', () => {
    const wrapper = buildDoorWrapper(DoubleSlideDoor, mainDoor, RomiCore.DoorMode.MODE_MOVING);
    const singleDoorSVGLine = getSingleDoorSVGLine(wrapper);
    expect(singleDoorSVGLine.hasClass(/(makeStyles)-(doorProcess)-(\d+)/)).toBe(true);
    wrapper.unmount();
  });
});
