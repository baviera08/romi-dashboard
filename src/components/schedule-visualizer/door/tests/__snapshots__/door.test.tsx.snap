// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Doors renders correctly 1`] = `
<svg>
  <g
    onClick={[Function]}
  >
    <line
      className="makeStyles-doorMarker-8 makeStyles-door-9 makeStyles-doorOpen-11"
      x1={8.2}
      x2={7.85}
      y1={5.5}
      y2={6.2}
    />
  </g>
  <g
    onClick={[Function]}
  >
    <line
      className="makeStyles-doorMarker-8 makeStyles-door-9 makeStyles-doorTransparent-14"
      x1={8.2}
      x2={7.85}
      y1={5.5}
      y2={6.2}
    />
  </g>
</svg>
`;

exports[`Doors renders correctly 2`] = `
<svg>
  <g
    onClick={[Function]}
  >
    <line
      className="makeStyles-doorMarker-8 makeStyles-door-9 makeStyles-doorOpen-11"
      x1={8.2}
      x2={8.024999999999999}
      y1={5.5}
      y2={5.85}
    />
  </g>
  <g
    onClick={[Function]}
  >
    <line
      className="makeStyles-doorMarker-8 makeStyles-door-9 makeStyles-doorTransparent-14"
      x1={8.2}
      x2={8.024999999999999}
      y1={5.5}
      y2={5.85}
    />
  </g>
  <g
    onClick={[Function]}
  >
    <line
      className="makeStyles-doorMarker-8 makeStyles-door-9 makeStyles-doorOpen-11"
      x1={8.024999999999999}
      x2={7.85}
      y1={5.85}
      y2={6.2}
    />
  </g>
  <g
    onClick={[Function]}
  >
    <line
      className="makeStyles-doorMarker-8 makeStyles-door-9 makeStyles-doorTransparent-14"
      x1={8.024999999999999}
      x2={7.85}
      y1={5.85}
      y2={6.2}
    />
  </g>
</svg>
`;

exports[`Doors renders correctly 3`] = `
<svg>
  <g
    onClick={[Function]}
  >
    <line
      className="makeStyles-doorMarker-8 makeStyles-door-9 makeStyles-doorOpen-11"
      x1={8.2}
      x2={7.85}
      y1={5.5}
      y2={6.2}
    />
  </g>
  <g
    onClick={[Function]}
  >
    <line
      className="makeStyles-doorMarker-8 makeStyles-door-9 makeStyles-doorTransparent-14"
      x1={8.2}
      x2={7.85}
      y1={5.5}
      y2={6.2}
    />
  </g>
</svg>
`;

exports[`Doors renders correctly 4`] = `
<svg>
  <g
    onClick={[Function]}
  >
    <line
      className="makeStyles-doorMarker-8 makeStyles-door-9 makeStyles-doorOpen-11"
      x1={8.2}
      x2={8.024999999999999}
      y1={5.5}
      y2={5.85}
    />
  </g>
  <g
    onClick={[Function]}
  >
    <line
      className="makeStyles-doorMarker-8 makeStyles-door-9 makeStyles-doorTransparent-14"
      x1={8.2}
      x2={8.024999999999999}
      y1={5.5}
      y2={5.85}
    />
  </g>
  <g
    onClick={[Function]}
  >
    <line
      className="makeStyles-doorMarker-8 makeStyles-door-9 makeStyles-doorOpen-11"
      x1={8.024999999999999}
      x2={7.85}
      y1={5.85}
      y2={6.2}
    />
  </g>
  <g
    onClick={[Function]}
  >
    <line
      className="makeStyles-doorMarker-8 makeStyles-door-9 makeStyles-doorTransparent-14"
      x1={8.024999999999999}
      x2={7.85}
      y1={5.85}
      y2={6.2}
    />
  </g>
</svg>
`;
